### DISCUSSION TITLE
BadTcpEndpointUrlInvalid error after executing set_security_string

### DISCUSSION THREAD

**Discussion by MaxWagels** [FreeOpcUa/opcua-asyncio#812] [Q&A]
Hi,
I want to connect to an encrypted OPC UA production server. To establish a connection I use the function `set_security_string`. I create the necessary certificate and key as described in the [example](https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/generate_certificate.sh).
The `serverURL`, `username`, `password`, `SecurityPolicy` and `SecurityMode` are all correct. 
If I now execute the script, the error shown as `set_security_string` is executed.
The OPC UA server is set to accept every incoming certificate automatically. However, the certificate I created is not displayed.
I can't figure out the error and when I try to connect to the Prosys Simulation Server with the same settings, it works fine.
For the creation of the certificate I have chosen
`subjectAltName = URI:urn:freeopcua:client`
and for the client
`client.application_uri = "urn:freeopcua:client"`
What am I doing wrong?
Thanks for your help

![Error1](https://user-images.githubusercontent.com/64021553/154460115-b0a89dfb-5734-49f1-90e8-af7fc4cb8669.png)
![Error2](https://user-images.githubusercontent.com/64021553/154460127-f9820bf0-9287-46df-80d4-3ada3868ee52.png)

**Comment by AndreasHeine** [comment]
could you post a simplified example?

**Reply by MaxWagels** [reply]
```
import asyncio
import logging
from asyncua import Client

async def main():
    url = "opc.tcp://opcuaserver.com:48010" # add server url here
    username = "TestUser"   # set the username to access the server
    userpassword = "12345"  # set the password to access the server
    try:
        client=Client(url=url, timeout=30) # set timeout to 30 seconds
        client.set_user(username) 
        client.set_password(userpassword) 
        
        client.application_uri = "urn:freeopcua:client" # Should match in your certificate
        await client.set_security_string("Basic256Sha256,SignAndEncrypt,C:\\Users\\Name\\Desktop\\my_cert.der,C:\\Users\\Name\\Desktop\\my_private_key.pem")
        
        await client.connect() # connect to opc ua server
        print("Root children are", await client.nodes.root.get_children())
    finally:
        await client.disconnect()

if __name__ == "__main__":
    logging.basicConfig(level=logging.WARN)
    asyncio.run(main())
```

**Reply by MaxWagels** [reply]
Tomorrow I can work with the machine again, do you have an idea what I can do to prevent the error or how I can find out exactly what the problem is?
